phenol-analysis/src/main/java/org/monarchinitiative/phenol/analysis/GeneAnnotations.java

```diff
@@ -24,7 +24,7 @@ public static GeneAnnotations of(TermId annotatedGene, List<TermAnnotation> anno
 
     /**
      * @param annotatedGene name of the gene or other item being annotated
-     * @param annotations
+     * @param annotations Gene Ontology annotations for the annotatedGene
      */
     private GeneAnnotations(TermId annotatedGene, List<TermAnnotation> annotations) {
```

Comment:
Is `TermId` still used for gene identifiers? Wouldn't these be better served using @ielis new [GeneIdentifier](https://github.com/monarch-initiative/phenol/pull/352/files#diff-fcd828f8319a2ce7c9517a8bbc60e69b96de0ab32c3aa451bf55c88fcc93b020) class? Same for uses of `TermId` in the `AnnotationContainer`.

Reply:
I agree with this and experimented with using AccessionNumbers but did not have enough time to figure out how to get the generics to work to allow TermId or AccessionNumber. It would be better to have AssociationContainer<T>, with GeneContainer extends AssociationCounter<AccessionNumber>. Possibly this could go in the next release...

Reply:
This is the current code signature:
```java
// T should be representing the generic difference, which is either a TermAnnotation or a GoGaf22Annotation 
public interface AssociationContainer<T> {

// here 'TermId' is always representing a gene identifier CURIE, so might be better typed as a GeneIdentifier
  Map<TermId, List<T>> getOntologyTermToDomainItemsMap();

  Map<T, DirectAndIndirectTermAnnotations> getAssociationMap(Set<TermId> annotatedItemTermIds);

 // this is also always referring to a GeneIdentifier, not a T (this is confusing because T is currently always a TermId)
  Set<T> getAllAnnotatedGenes();
}

public class TermAssociationContainer implements AssociationContainer<TermId> {}
// internally the TermId here is always referring to a gene identifier/ gene symbol so maybe this should also be a GeneIdentifier
public class GoAssociationContainer implements AssociationContainer<TermId> {}
// shouldn't this be 
public class GoAssociationContainer implements AssociationContainer<GeneIdentifier> {}
```

which I think this ought to look like this:

```java
// T should be representing the generic difference, which is either a TermAnnotation or a GoGaf22Annotation 
public interface AssociationContainer<T> {

  // Now need to ask, is this the correct method name?
// both implementations have this signature:
//  Map<TermId, List<TermId>> getOntologyTermToDomainItemsMap() {
  Map<GeneIdentifier, List<T>> getOntologyTermToDomainItemsMap();

// Or is this also actually referring to GeneIdentifier? I think it is.
  Map<T, DirectAndIndirectTermAnnotations> getAssociationMap(Set<GeneIdentifier> annotatedItemTermIds);

  Set<GeneIdentifier> getAllAnnotatedGenes();
}

public class TermAssociationContainer implements AssociationContainer<TermAnnotation> {}

public class GoAssociationContainer implements AssociationContainer<GoGaf22Annotation> {}
```

But actually, the `T` type is never exposed in the interface (trying to judge from the use of these in the classes). So this means we can remove the generic <T> because it is actually just an implementation detail - the classes are almost identical copy-paste code. 

```java
// T should be representing the generic difference, which is either a TermAnnotation or a GoGaf22Annotation but this is never exposed through the interface, so remove it.
public interface GeneAssociationContainer {

  // Now need to ask, is this the correct method name?
  // both implementations have this signature:
  //  Map<TermId, List<TermId>> getOntologyTermToDomainItemsMap() {
  Map<GeneIdentifier, List<TermId>> getGeneIdentifierToDomainItems();

  Map<GeneIdentifier, DirectAndIndirectTermAnnotations> getAssociationMap(Set<GeneIdentifier> annotatedGenes);

  Set<GeneIdentifier> getAllAnnotatedGenes();
}

public class TermAssociationContainer implements GeneAssociationContainer {}

public class GoAssociationContainer implements GeneAssociationContainer {}
```

Would that all make sense with what you're trying to do? Are these functionally equivalent, only differing in the data sources? Does this now prove my point about the lack of strong typing with `TermId` being confusing? If I've got the idea behind this completely wrong, I'm at least unassailably correct on the confusion point :)

Reply:
Looks like it's possible to go further as both classes require a parser to read the file and produce a set of `TermAnnotation` (`GoGaf22Annotation` is a `TermAnnotation`) s in the end it might make sense to have one concrete `GeneAssociationContainer` and two parsers/readers to extract the data from different files.

Reply:
Possibly, but the two kinds of annotation of very different and I doubt there would ever be use for polymorphism and also I doubt there woould be much code duplication if we did not do this.

Reply:
Yes, this might be a good plan. I'd also have a look at how `GeneAnnotations` now fits into the new worldview as that is essentially modelling a Map.Entry<TermId, List<TermAnnotation>> where the key should actually be a `GeneIdentifier`.

Reply:
Hi @pnrobinson , @julesjacobsen , after looking more closely at the containers mentioned above, it's hard for me to see any useful abstraction. So I'd leave it be for now.

Regarding `GeneAnnotations` & `GeneIdentifier` - I'm not sure I understand the usage of the `GeneAnnotations` in the analysis. The `GeneAnnotations` constructor takes `TermId annotatedGene`. However, at the only place where the constructor is being invoked (here), `TermAnnotation.itemId()` (**not** `TermAnotation.id()`) is used. I'm not sure `itemId` is gene identifier, and it is not possible to create `GeneIdentifier` out of the data.

So, I don't know. If it is not a must, maybe we can leave it like it is for now.